storage/maria/ma_checkpoint.c

```diff
@@ -811,7 +811,6 @@ static int collect_tables(LEX_STRING *str, LSN checkpoint_start_log_horizon)
   }
 
   DBUG_ASSERT(i == nb);
-  mysql_mutex_unlock(&THR_LOCK_maria);
```

Comment:
All this trouble with ```share->in_checkpoint``` was done to avoid holding THR_LOCK_mutex while executing following code. Which contains a bunch of heavy operations like disk IO, other mutex locks, etc. 

By default this job is ran by background thread every 30 seconds. Which means every 30 seconds you may see scalability dips for certain loads.

I believe the simplest fix would be to remove that ```goto err``` and reset ```share->in_checkpoint``` for all tables directly under that if; and return error.